mixer/adapter/cloudwatch/logHandler.go

```diff
@@ -0,0 +1,158 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY Type, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cloudwatch
+
+import (
+	"html/template"
+	"net"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/cloudwatchlogs"
+	multierror "github.com/hashicorp/go-multierror"
+
+	"istio.io/istio/mixer/pkg/pool"
+	"istio.io/istio/mixer/template/logentry"
+)
+
+const (
+	// Cloudwatchlogs enforced limit on number of logentries per request
+	// https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_PutLogEvents.html
+	batchcount = 10000
+
+	// A default template for log entries will be used if operator doesn't supply one
+	defaultTemplate = `{{or (.source_ip) "-"}} - {{or (.source_user) "-"}} ` +
+		`[{{or (.timestamp.Format "2006-01-02T15:04:05Z07:00") "-"}}] "{{or (.method) "-"}} {{or (.url) "-"}} ` +
+		`{{or (.protocol) "-"}}" {{or (.response_code) "-"}} {{or (.response_size) "-"}}`
+)
+
+func (h *handler) generateLogEntryData(insts []*logentry.Instance) []*cloudwatchlogs.InputLogEvent {
+	logentryData := make([]*cloudwatchlogs.InputLogEvent, 0, len(insts))
+
+	for _, inst := range insts {
+		// cloudwatchlogs accepts unix timestamp in milliseconds
+		timestamp := inst.Timestamp.UnixNano() / int64(time.Millisecond)
+
+		message, err := getMessageFromVariables(h, inst)
+		if err != nil {
+			h.env.Logger().Errorf("failed to get message for instance: %s. %v", inst.Name, err)
+			continue
+		}
+
+		logEvent := cloudwatchlogs.InputLogEvent{
+			Message:   aws.String(message),
+			Timestamp: aws.Int64(timestamp),
+		}
+
+		logentryData = append(logentryData, &logEvent)
+	}
+	return logentryData
+}
+
+func getMessageFromVariables(h *handler, inst *logentry.Instance) (string, error) {
+	tmplString := h.cfg.GetLogs()[inst.Name].PayloadTemplate
+	tmpl, err := template.New(inst.Name).Parse(tmplString)
+	if err != nil {
+		return "", h.env.Logger().Errorf("failed to evaluate template for log instance: %s, skipping: %v", inst, err)
+	}
+
+	buf := pool.GetBuffer()
+
+	data := inst.Variables
+	data["timestamp"] = inst.Timestamp
+	data["severity"] = inst.Severity
+
+	ipval, ok := inst.Variables["source_ip"].([]byte)
+	if ok {
+		inst.Variables["source_ip"] = net.IP(ipval).String()
+	}
+
+	if err := tmpl.Execute(buf, data); err != nil {
+		pool.PutBuffer(buf)
+		return "", h.env.Logger().Errorf("failed to execute template for log '%s': %v", inst.Name, err)
+	}
+
+	payload := buf.String()
+	pool.PutBuffer(buf)
+	return payload, nil
+}
+
+func (h *handler) sendLogEntriesToCloudWatch(logentryData []*cloudwatchlogs.InputLogEvent) (int, error) {
+	cloudWatchCallCount := 0
+	var multiError *multierror.Error
+
+	nextSequenceToken, err := getNextSequenceToken(h)
+	if err != nil {
+		h.env.Logger().Errorf("could not get next upload sequence token: %v", err)
```

Comment:
do we care about logging something about dropped messages ? or retrying once or twice here?

Reply:
changed the error message to also log upload failure